server/service/service_osquery.go

```diff
@@ -446,15 +447,79 @@ func (svc *Service) detailQueriesForHost(ctx context.Context, host *fleet.Host)
 }
 
 func (svc *Service) shouldUpdate(lastUpdated time.Time, interval time.Duration, hostID uint) bool {
-	var jitter time.Duration
-	if svc.config.Osquery.MaxJitterPercent > 0 {
-		maxJitter := int64(svc.config.Osquery.MaxJitterPercent) * int64(interval) / 100.0
-		jitter = time.Duration((int64(hostID) + svc.getJitterSeed()) % maxJitter)
+	svc.jitterMu.Lock()
+	defer svc.jitterMu.Unlock()
+
+	if svc.jitterH[interval] == nil {
+		svc.jitterH[interval] = &jitterHashTable{
+			maxCapacity: 1,
+			bucketCount: int(int64(svc.config.Osquery.MaxJitterPercent) * int64(interval.Minutes()) / 100.0),
+			buckets:     make(map[int]int),
+			cache:       make(map[uint]time.Duration),
+		}
+		if svc.jitterH[interval].bucketCount == 0 {
+			svc.jitterH[interval].bucketCount++
+		}
+		level.Debug(svc.logger).Log("jitter", "created", "bucketCount", svc.jitterH[interval].bucketCount)
 	}
+
+	jitter := svc.jitterH[interval].jitterForHost(hostID)
 	cutoff := svc.clock.Now().Add(-(interval + jitter))
 	return lastUpdated.Before(cutoff)
 }
 
+type jitterHashTable struct {
+	mu          sync.Mutex
+	maxCapacity int
+	bucketCount int
+	buckets     map[int]int
+	cache       map[uint]time.Duration
+}
+
+func (jh *jitterHashTable) jitterForHost(hostID uint) time.Duration {
+	jh.mu.Lock()
+	if jitter, ok := jh.cache[hostID]; ok {
+		jh.mu.Unlock()
+		return jitter
+	}
+
+	possibleBucket := int(hostID) % jh.bucketCount
+	if jh.buckets[possibleBucket] < jh.maxCapacity {
+		jh.buckets[possibleBucket]++
+		jitter := time.Duration(possibleBucket) * time.Minute
+		jh.cache[hostID] = jitter
+
+		jh.mu.Unlock()
+		return jitter
+	}
```

Comment:
This is the exact same code as inside the `for` loop below AFAICS, so you could just use the for loop and iterate from `i := 0; i < jh.bucketCount; i++`, that should make a full loop around all possible buckets (the `possibleBucket` variable starting at `hostID` and incrementing by `i`).

Reply:
Ah, great catch! Much simpler.